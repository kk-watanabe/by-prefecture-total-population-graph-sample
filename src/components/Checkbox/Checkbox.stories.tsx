import { Checkbox } from "./Checkbox";

import type { ComponentStory, ComponentMeta } from "@storybook/react";

export default {
  title: "Checkbox",
  component: Checkbox,
  argTypes: {
    onChange: { action: "changed" },
  },
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = (args) => <Checkbox {...args} />;

export const Default = Template.bind({});
Default.args = {
  id: "checkbox-id1",
  label: "Checkbox label",
  value: 1,
  checked: false,
  colorCode: "red",
};

export const Checked = Template.bind({});
Checked.args = { ...Default.args, checked: true };
