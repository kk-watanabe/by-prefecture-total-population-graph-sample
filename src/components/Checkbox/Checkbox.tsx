import { ChangeEvent, useState, useCallback } from "react";

import styles from "./Styles.module.css";

export interface CheckboxProps {
  id: string;
  label: string;
  value: string | number;
  checked?: boolean;
  colorCode: string;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
}

export const Checkbox = (props: CheckboxProps) => {
  const { onChange } = props;
  const [isChecked, setIsChecked] = useState(props.checked);
  const checkedBoxClassName = isChecked ? styles["checkbox__box--checked"] : "";
  const boxClasses = [styles.checkbox__box, checkedBoxClassName].join(" ");
  const checkedBoxStyle = {
    backgroundColor: isChecked ? props.colorCode : "",
    borderColor: isChecked ? props.colorCode : "",
  };

  const checkedLabelClassName = isChecked ? styles["checkbox__label--checked"] : "";
  const labelClasses = [styles.checkbox__label, checkedLabelClassName].join(" ");

  const handleChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      if (onChange) {
        setIsChecked(event.target.checked);
        onChange(event);
      }
    },
    [onChange]
  );

  return (
    <label className={styles.checkbox} htmlFor={props.id}>
      <input
        type="checkbox"
        value={props.value}
        id={props.id}
        defaultChecked={isChecked}
        onChange={handleChange}
        className={styles.checkbox__input}
      />
      <span className={boxClasses} style={checkedBoxStyle}>
        <span className={styles.checkbox__check} />
      </span>
      <span className={labelClasses}>{props.label}</span>
    </label>
  );
};
