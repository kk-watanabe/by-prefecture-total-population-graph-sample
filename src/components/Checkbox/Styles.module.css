.checkbox {
  position: relative;
  display: inline-flex;
  align-items: center;
  cursor: pointer;

  gap: 8px;
}

.checkbox__input {
  position: absolute;
  top: 0;
  right: 0;
  width: 1px;
  height: 1px;
  opacity: 0;
}

.checkbox__box {
  position: relative;
  display: block;
  width: 20px;
  height: 20px;
  border: 1px solid var(--color-gray300);
  border-radius: 4px;
  background-color: var(--color-white);
  transition: border-color 0.1s ease;
}

.checkbox__check {
  position: absolute;
  top: 4px;
  left: 4px;
  width: 10px;
  height: 6px;
  border-bottom: 2px solid var(--color-white);
  border-left: 2px solid var(--color-white);
  transform: rotate(-45deg);
}

.checkbox__label {
  display: block;
  padding-top: 2px;
  color: var(--color-gray700);
  font-size: var(--fs-sm);
}

.checkbox__label--checked {
  color: var(--color-gray900);
}

/* hover */
.checkbox:hover .checkbox__box {
  border-color: var(--color-gray500);
}
