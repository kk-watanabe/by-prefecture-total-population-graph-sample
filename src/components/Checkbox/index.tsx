/* istanbul ignore next */
export * from "./Checkbox";
