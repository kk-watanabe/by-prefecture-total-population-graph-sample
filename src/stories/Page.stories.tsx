import React from "react";

import { Page } from "./Page";

import type { ComponentStory, ComponentMeta } from "@storybook/react";

export default {
  title: "Example/Page",
  component: Page,
  parameters: {
    // More on Story layout: https://storybook.js.org/docs/react/configure/story-layout
    layout: "fullscreen",
  },
} as ComponentMeta<typeof Page>;

const Template: ComponentStory<typeof Page> = (args) => <Page {...args} />;

export const LoggedOut = Template.bind({});

export const LoggedIn = Template.bind({});
