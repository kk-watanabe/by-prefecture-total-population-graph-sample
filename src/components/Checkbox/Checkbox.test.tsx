import { composeStories } from "@storybook/testing-react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";

import * as stories from "./Checkbox.stories";

const { Default, Checked } = composeStories(stories);

describe("snapshots", () => {
  it("Readner Default components", () => {
    const { container } = render(<Default />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Readner Checked components", () => {
    const { container } = render(<Checked />);

    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("label click", async () => {
  it("change checked components", async () => {
    render(<Default />);

    const checkbox = screen.getByRole("checkbox");

    expect(checkbox).not.toBeChecked();
    fireEvent.click(checkbox, true);

    await waitFor(() => {
      expect(checkbox).toBeChecked();
    });
  });

  it("change not checked components", async () => {
    render(<Checked />);

    const checkbox = screen.getByRole("checkbox");

    expect(checkbox).toBeChecked();
    fireEvent.click(checkbox, false);

    await waitFor(() => {
      expect(checkbox).not.toBeChecked();
    });
  });
});
