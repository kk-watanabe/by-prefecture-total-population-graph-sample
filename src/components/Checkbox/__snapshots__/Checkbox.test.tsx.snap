// Vitest Snapshot v1

exports[`snapshots > Readner Checked components 1`] = `
<label
  class="_checkbox_qo6px_1"
  for="checkbox-id1"
>
  <input
    checked=""
    class="_checkbox__input_qo6px_10"
    id="checkbox-id1"
    type="checkbox"
    value="1"
  />
  <span
    class="_checkbox__box_qo6px_19 "
    style="background-color: red; border-color: red;"
  >
    <span
      class="_checkbox__check_qo6px_29"
    />
  </span>
  <span
    class="_checkbox__label_qo6px_40 _checkbox__label--checked_qo6px_47"
  >
    Checkbox label
  </span>
</label>
`;

exports[`snapshots > Readner Default components 1`] = `
<label
  class="_checkbox_qo6px_1"
  for="checkbox-id1"
>
  <input
    class="_checkbox__input_qo6px_10"
    id="checkbox-id1"
    type="checkbox"
    value="1"
  />
  <span
    class="_checkbox__box_qo6px_19 "
  >
    <span
      class="_checkbox__check_qo6px_29"
    />
  </span>
  <span
    class="_checkbox__label_qo6px_40 "
  >
    Checkbox label
  </span>
</label>
`;
